Stanza - Enhanced Testing / Enable Component

djaglowski opened the pull request:
**Description:** 
Adds significant testing to the stanza receiver, including:
- Benchmarks for conversion from stanza log format
- End to end testing
- Testbed performance tests.

Also enables stanza as a component in the contrib collector.

djaglowski committed:
Add stanza logs receiver

djaglowski committed:
Cleanup dependencies

djaglowski committed:
Exclude other component dependencies

djaglowski committed:
Merged master

djaglowski committed:
Merged master

djaglowski committed:
Fix merge issues

djaglowski committed:
Removed testbed tests

djaglowski committed:
Go tidy

djaglowski committed:
Merge branch 'master' into stanzareceiver

tigrannajaryan commented on receiver/stanzareceiver/receiver_test.go:
```diff
@@ -72,3 +80,107 @@ func TestHandleConsumeError(t *testing.T) {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE, 0666)
+	require.NoError(b, err)
+	for i := 0; i < b.N; i++ {
+		file.WriteString("testlog\n")
```
I am not sure I fully understand what we benchmark. Looks like we are writing N lines to a file and then read them all. Does this benchmark measure the time it takes to write and read a single line on average?

djaglowski replied:
It should be measuring how long it takes to read and transform a single line on average, but I was missing the call to `convert`, which I'll add momentarily.

As you said, we're writing N lines to a file, but then we are:
- resetting the benchmark timer to exclude time spent writing
- starting the stanza pipeline, which begins consuming the file
- as logs are tailed, convert them from stanza format to `pdata.Logs`

```
// Run the actual benchmark
b.ResetTimer()
require.NoError(b, pl.Start())
for i := 0; i < b.N; i++ {
	convert(<-emitter.logChan)
}
```

To be clear, I think Ben pointed out that there is room for improvement in terms of efficiency here. I think the current implementation is acceptable though and wanted to optimize for performance in a future PR.

tigrannajaryan commented on receiver/stanzareceiver/receiver_test.go:
```diff
@@ -72,3 +80,107 @@ func TestHandleConsumeError(t *testing.T) {
+	}
+}
+
+func BenchmarkPipelineComplex(b *testing.B) {
```
```suggestion
func BenchmarkParseAndMap(b *testing.B) {
```

tigrannajaryan commented on receiver/stanzareceiver/receiver_test.go:
```diff
@@ -72,3 +80,107 @@ func TestHandleConsumeError(t *testing.T) {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE, 0666)
+	require.NoError(b, err)
+	for i := 0; i < b.N; i++ {
+		file.WriteString("10.33.121.119 - - [11/Aug/2020:00:00:00 -0400] \"GET /index.html HTTP/1.1\" 404 761\n")
```
To make test data a bit more realistic can we use `i` in the string?

tigrannajaryan commented:
>Also enables stanza as a component in the contrib collector.

@djaglowski can we skip this part for now, while I will figure out the best way to enable it? The rest looks good to me.

djaglowski committed:
Update receiver/stanzareceiver/e2e_test.go

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

djaglowski committed:
Update receiver/stanzareceiver/receiver_test.go

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

djaglowski committed:
Clean up some test code per feedback

djaglowski committed:
Merge branch 'stanzareceiver' of github.com:observIQ/opentelemetry-collector-contrib into stanzareceiver

tigrannajaryan commented:
Can you also post the benchmark results in the commit message / PR description?

djaglowski commented on receiver/stanzareceiver/e2e_test.go:
```diff
@@ -0,0 +1,153 @@
+func newRotatingLogger(t *testing.T, tempDir string, maxLines, maxBackups int, copyTruncate bool) *log.Logger {
+	path := filepath.Join(tempDir, "test.log")
+	rotator := &nanojack.Logger{
+		Filename:     path,
```
Similar idea, but the extensions are timestamped with `2006-01-02T15-04-05.000000000` format, such as `test.log.2020-10-20T15-04-05.000000000`.

I can (and would like to) enhance `nanojack` to allow for various backup file naming conventions, but I will point out that what I think we're interested in here ultimately boils down to a question of whether or not `stanza` can correctly detect new files and files with previously seen names that have changed. I believe we cover these quite well in stanza unit tests, but I can appreciate that it would be worth covering here as well. I think enhancing `nanojack` to support `.n` extension should do both.

I will take a look at `nanojack` today.

djaglowski replied:
Tests have now been updated to include exactly this rotation pattern.

tigrannajaryan replied:
Can you please add comments to explain this in the code?

djaglowski committed:
Back out making stanza a default component

djaglowski committed:
End to end static file compares exact log messages

djaglowski committed:
Enhance end to end test to also validate each log line

djaglowski commented on receiver/stanzareceiver/e2e_test.go:
```diff
@@ -0,0 +1,153 @@
+		time.Sleep(time.Millisecond)
+	}
+
+	require.Eventually(t, expectNLogs(sink, numLogs), 2*time.Second, time.Millisecond)
```
Done

djaglowski committed:
Test sequential file rotation

djaglowski commented:
Benchmark results:
```
> (cd receiver/stanzareceiver/ && make benchmark)
go test -bench=. -run=notests github.com/open-telemetry/opentelemetry-collector-contrib/receiver/stanzareceiver
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/stanzareceiver
BenchmarkConvertSimple-8         1000000              1076 ns/op
BenchmarkConvertComplex-8          60814             16710 ns/op
BenchmarkReadLines-8              290959              4249 ns/op
BenchmarkParseAndMap-8             57204             20619 ns/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/receiver/stanzareceiver       21.395s
```

Times are up about 50% after merging master. I'm going to look into this further, but I'm still of the opinion that this is an acceptable amount of throughput for now.

@tigrannajaryan Thanks again for the review. I believe I've addressed all feedback. I will plan on optimizing against these benchmark in a near future PR, unless you are concerned with current results.

tigrannajaryan commented on receiver/stanzareceiver/receiver_test.go:
```diff
@@ -72,3 +80,107 @@ func TestHandleConsumeError(t *testing.T) {
 	require.Equal(t, 1, mockConsumer.rejected, "one log entry expected")
 }
+
+func BenchmarkReadLines(b *testing.B) {
```
Nit: `BenchmarkReadLine` since it shows the time for a single line not lines.

djaglowski replied:
Good call. Updated.

tigrannajaryan reviewed (approved):
@djaglowski LGTM, throughput is fine, future optimizations are welcome.

djaglowski committed:
Singularize benchmark name
